### ISSUE TITLE
Can not access Traffic Vault API when using SSL

### ISSUE THREAD

**Issue by riuzen** [apache/trafficcontrol#4138]
## I'm submitting a ...

- [ ] bug report
- [ ] new feature / enhancement request
- [ ] improvement request (usability, performance, tech debt, etc.)
- [ ] other 

## Traffic Control components affected ...

- [ ] CDN in a Box
- [ ] Documentation
- [ ] Grove
- [ ] Traffic Control Client
- [ ] Traffic Monitor
- [ ] Traffic Ops
- [ ] Traffic Ops ORT
- [ ] Traffic Portal
- [ ] Traffic Router
- [ ] Traffic Stats
- [x] Traffic Vault
- [ ] unknown

## Current behavior:

I follow this guide https://traffic-control-cdn.readthedocs.io/en/latest/admin/traffic_vault.html to install traffic vault on 1 VM. When I enable SSL, this error appeared. I generate self-signed cert like in the guide.
```
[root@traffic-ops ~]# curl -v https://trafficvault.ethernet.id:8088/types/default/props
* About to connect() to trafficvault.ethernet.id port 8088 (#0)
*   Trying 192.168.30.107...
* Connected to trafficvault.ethernet.id (192.168.30.107) port 8088 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* NSS error -5938 (PR_END_OF_FILE_ERROR)
* Encountered end of file
* Closing connection 0
curl: (35) Encountered end of file
[root@traffic-ops ~]# 
```

## Minimal reproduction of the problem with instructions:

This is the riak error.log.

```
2019-11-21 14:02:32.296 [error] <0.91.0> Supervisor ssl_connection_sup had child undefined started with {tls_connection,start_link,undefined} at <0.6302.0> exit with reason no function clause matching ssl_cipher:hash_algorithm(8) line 1174 in context child_terminated
2019-11-21 14:02:32.296 [error] <0.345.0> application: mochiweb, "Accept failed error", "{'EXIT',\n    {{function_clause,\n         [{ssl_cipher,hash_algorithm,\"\\b\",\n              [{file,\"ssl_cipher.erl\"},{line,1174}]},\n          {ssl_handshake,'-dec_hello_extensions/2-blc$^0/1-1-',1,\n              [{file,\"ssl_handshake.erl\"},{line,1487}]},\n          {ssl_handshake,'-dec_hello_extensions/2-blc$^0/1-1-',1,\n              [{file,\"ssl_handshake.erl\"},{line,1488}]},\n          {ssl_handshake,dec_hello_extensions,2,\n              [{file,\"ssl_handshake.erl\"},{line,1487}]},\n          {tls_handshake,decode_handshake,3,\n              [{file,\"tls_handshake.erl\"},{line,236}]},\n          {tls_handshake,get_tls_handshake_aux,3,\n              [{file,\"tls_handshake.erl\"},{line,203}]},\n          {tls_connection,next_state,4,\n              [{file,\"tls_connection.erl\"},{line,2556}]},\n          {gen_fsm,handle_msg,7,[{file,\"gen_fsm.erl\"},{line,505}]}]},\n     {gen_fsm,sync_send_all_state_event,\n         [<0.6302.0>,{start,infinity},infinity]}}}"
2019-11-21 14:02:32.296 [error] <0.345.0> CRASH REPORT Process <0.345.0> with 0 neighbours exited with reason: {error,accept_failed} in mochiweb_acceptor:init/3 line 33
2019-11-21 14:02:32.296 [error] <0.335.0> {mochiweb_socket_server,320,{acceptor_error,{error,accept_failed}}}
2019-11-21 14:02:54.690 [error] <0.6337.0> gen_fsm <0.6337.0> in state hello terminated with reason: no function clause matching ssl_cipher:hash_algorithm(8) line 1174
2019-11-21 14:02:54.690 [error] <0.6337.0> CRASH REPORT Process <0.6337.0> with 0 neighbours exited with reason: no function clause matching ssl_cipher:hash_algorithm(8) line 1174 in gen_fsm:terminate/7 line 622
2019-11-21 14:02:54.690 [error] <0.91.0> Supervisor ssl_connection_sup had child undefined started with {tls_connection,start_link,undefined} at <0.6337.0> exit with reason no function clause matching ssl_cipher:hash_algorithm(8) line 1174 in context child_terminated
2019-11-21 14:02:54.691 [error] <0.347.0> application: mochiweb, "Accept failed error", "{'EXIT',\n    {{function_clause,\n         [{ssl_cipher,hash_algorithm,\"\\b\",\n              [{file,\"ssl_cipher.erl\"},{line,1174}]},\n          {ssl_handshake,'-dec_hello_extensions/2-blc$^0/1-1-',1,\n              [{file,\"ssl_handshake.erl\"},{line,1487}]},\n          {ssl_handshake,'-dec_hello_extensions/2-blc$^0/1-1-',1,\n              [{file,\"ssl_handshake.erl\"},{line,1488}]},\n          {ssl_handshake,dec_hello_extensions,2,\n              [{file,\"ssl_handshake.erl\"},{line,1487}]},\n          {tls_handshake,decode_handshake,3,\n              [{file,\"tls_handshake.erl\"},{line,236}]},\n          {tls_handshake,get_tls_handshake_aux,3,\n              [{file,\"tls_handshake.erl\"},{line,203}]},\n          {tls_connection,next_state,4,\n              [{file,\"tls_connection.erl\"},{line,2556}]},\n          {gen_fsm,handle_msg,7,[{file,\"gen_fsm.erl\"},{line,505}]}]},\n     {gen_fsm,sync_send_all_state_event,\n         [<0.6337.0>,{start,infinity},infinity]}}}"
2019-11-21 14:02:54.691 [error] <0.347.0> CRASH REPORT Process <0.347.0> with 0 neighbours exited with reason: {error,accept_failed} in mochiweb_acceptor:init/3 line 33
2019-11-21 14:02:54.691 [error] <0.335.0> {mochiweb_socket_server,320,{acceptor_error,{error,accept_failed}}}
```

**Comment by ocket8888** [comment]
What's the output of `ls /etc/riak/certs/` and `ls /etc/pki/tls/certs/`?

**Comment by riuzen** [comment]
This...

```
[root@atc-traffic-vault ~]# ll /etc/pki/tls/certs/
total 16
lrwxrwxrwx 1 root root   49 Nov 21 13:34 ca-bundle.crt -> /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
lrwxrwxrwx 1 root root   55 Nov 21 13:34 ca-bundle.trust.crt -> /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
-rw-r--r-- 1 root root 1285 Nov 21 12:54 ethernet-bundle.crt
-rwxr-xr-x 1 root root  610 Agu  9 08:38 make-dummy-cert
-rw-r--r-- 1 root root 2516 Agu  9 08:38 Makefile
-rwxr-xr-x 1 root root  829 Agu  9 08:38 renew-dummy-cert

[root@atc-traffic-vault ~]# ll /etc/riak/certs/
total 16
-rw-r--r-- 1 root root 1285 Nov 21 12:51 ethernet.id.crt
-rw-r--r-- 1 root root 1675 Nov 21 12:14 ethernet.id.key
-rw-r--r-- 1 root root 1976 Nov 21 13:23 tes.crt
-rw-r--r-- 1 root root 3243 Nov 21 13:23 tes.key

[root@atc-traffic-vault ~]# cat /etc/riak/riak.conf |grep ssl
ssl.certfile = /etc/riak/certs/ethernet.id.crt
ssl.keyfile = /etc/riak/certs/ethernet.id.key
ssl.cacertfile = /etc/pki/tls/certs/ca-bundle.crt
```

**Comment by riupie** [comment]
Update:

Request to API success when i request from ubuntu 16 client, but it get error when make  request from ubuntu 18 or centos 7

**Comment by ocket8888** [comment]
Hm. Okay, are those certs (`/etc/riak/certs/ethernet.id.crt`/`/etc/riak/certs/ethernet.id.key`) self-signed? If so, can you try adding the `-k` option to your `curl` command?

**Comment by riuzen** [comment]
Can be accessed when requested from ubuntu 16.04 client.
```
root@vault-ubuntu:~# curl -vvv https://admin:yasmin88@trafficvault.ethernet.id:8088/types/default/props -k
*   Trying 192.168.30.107...
* Connected to trafficvault.ethernet.id (192.168.30.107) port 8088 (#0)
* found 148 certificates in /etc/ssl/certs/ca-certificates.crt
* found 592 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / DHE_RSA_AES_128_CBC_SHA1
* 	 server certificate verification SKIPPED
* 	 server certificate status verification SKIPPED
* 	 common name: *.ethernet.id (matched)
* 	 server certificate expiration date OK
* 	 server certificate activation date OK
* 	 certificate public key: RSA
* 	 certificate version: #1
* 	 subject: C=ID,ST=WestJava,L=Bogor,O=Btech,OU=Btech,CN=*.ethernet.id,EMAIL=ops@btech.id
* 	 start date: Thu, 21 Nov 2019 05:51:29 GMT
* 	 expire date: Fri, 20 Nov 2020 05:51:29 GMT
* 	 issuer: C=ID,ST=WestJava,L=Bogor,O=Btech,OU=Btech,CN=*.ethernet.id,EMAIL=ops@btech.id
* 	 compression: NULL
* ALPN, server did not agree to a protocol
* Server auth using Basic with user 'admin'
> GET /types/default/props HTTP/1.1
> Host: trafficvault.ethernet.id:8088
> Authorization: Basic YWRtaW46eWFzbWluODg=
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Vary: Accept-Encoding
< Server: MochiWeb/1.1 WebMachine/1.10.5 (jokes are better explained)
< Date: Fri, 22 Nov 2019 00:49:47 GMT
< Content-Type: application/json
< Content-Length: 428
< 
* Connection #0 to host trafficvault.ethernet.id left intact
{"props":{"allow_mult":false,"basic_quorum":false,"big_vclock":50,"chash_keyfun":{"mod":"riak_core_util","fun":"chash_std_keyfun"},"dvv_enabled":false,"dw":"quorum","last_write_wins":false,"linkfun":{"mod":"riak_kv_wm_link_walker","fun":"mapreduce_linkfun"},"n_val":3,"notfound_ok":true,"old_vclock":86400,"postcommit":[],"pr":0,"precommit":[],"pw":0,"r":"quorum","rw":"quorum","small_vclock":50,"w":"quorum","young_vclock":20}}
```

But error from another client OS like centos 7 and ubuntu 18. Hmm...did I miss something?

**Comment by ocket8888** [comment]
so you're saying it worked when you curled from an Ubuntu machine but non from some other distribution? What distro did you try first?

**Comment by riuzen** [comment]
I curled from centos 7 (from riak server itself) in my first try, then from traffic ops server (also cen7). Then, from my laptop (ubuntu18) -->all not work.
Then, i create a new VM ubuntu16,actually i want to try install riak server on this vm. It works perfectly. Then I also try to curled my centos riak server from this ubuntu and it also works.

TL;DR:
- Riak server works when I installed on Ubuntu 16.04 but failed in centos 7.
- I can curled my centos Riak Server from that Ubuntu 16.04, but failed when curled from other distributions.

**Comment by riuzen** [comment]
So, I also try to install CIAB for testing. Then I curled between my trafficops container to trafficvault container. And give me same result.
```
[mynode@ag-atc ~]$ docker exec -it 88b029985f22 curl -v https://trafficvault:8088/types/default/props
* About to connect() to trafficvault port 8088 (#0)
*   Trying 172.25.0.13...
* Connected to trafficvault (172.25.0.13) port 8088 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* NSS error -5938 (PR_END_OF_FILE_ERROR)
* Encountered end of file
* Closing connection 0
curl: (35) Encountered end of file
```

So, I just wondering..if I cant access its API from centos, how I could do this step?:
```
wget https://raw.githubusercontent.com/apache/trafficcontrol/master/traffic_ops/app/conf/misc/riak_search/sslkeys.xml
curl -kvsX PUT "https://admin:pass@trafficvault.infra.ciab.test:8088/search/schema/sslkeys" -H "Content-Type: application/xml" -d @sslkeys.xml
```

**Comment by ocket8888** [comment]
Well as far as CDN-in-a-Box goes, Riak is listening on port 8087 for unsecured connections, not port 8088 for HTTPS connections.

Truthfully, at this point I'm out of ideas. Hoping someone else can jump in on this.

**Comment by jhg03a** [comment]
I would double check the versions of curl you using to test with.  I expect them to be very different between ubuntu and centos due to how their upstream projects are managed.  From what I gather that error is related to the inability for curl to properly negotiate a protocol and cipher.

I'm not sure I follow entirely what you're doing with vault and ubuntu as that's an unsupported configuration.

I recommend adding the following line to the bottom of your riak.conf, restarting, and trying again [per the riak docs](https://docs.riak.com/riak/kv/latest/using/security/basics/index.html#tls-settings):
```
tls_protocols.tlsv1 = on
```

**Comment by riuzen** [comment]
> I would double check the versions of curl you using to test with. I expect them to be very different between ubuntu and centos due to how their upstream projects are managed. From what I gather that error is related to the inability for curl to properly negotiate a protocol and cipher.
> 
> I'm not sure I follow entirely what you're doing with vault and ubuntu as that's an unsupported configuration.
> 
> I recommend adding the following line to the bottom of your riak.conf, restarting, and trying again [per the riak docs](https://docs.riak.com/riak/kv/latest/using/security/basics/index.html#tls-settings):
> 
> ```
> tls_protocols.tlsv1 = on
> ```

I already add this line `tls_protocols.tlsv1 = on` same like in the guide. Yeah I know there is difference version curl in  both OS. Maybe we need add more detail about traffic vault guide? So it the vault could be access by difference client OS.

**Comment by midchildan** [comment]
I’ve also run into this problem while setting up Traffic Control. AFAICT, the main cause of this issue stems from the latest binary release of Riak KV, which ships with a buggy version of erlang (http://erlang.org/pipermail/erlang-questions/2015-June/084868.html). 

> Well as far as CDN-in-a-Box goes, Riak is listening on port 8087 for unsecured connections, not port 8088 for HTTPS connections.

I've made numerous attempts to get this working, but Riak also refuses to  accept protobuf connections (port 8087). Since it complains about not being able to start a SSL session when SSL is disabled, I think we have the same problem when using protobuf as well.

Here’s some log output that I’ve gathered:

### SSL enabled (`riak-admin security enable`)
#### Traffic Ops
```
ERROR: log.go:172: 2019-12-06T14:57:14.515182211+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: read tcp <TRAFFIC_OPS_IP>:56544-><TRAFFIC_VAULT_IP>:8087: i/o timeout
ERROR: log.go:172: 2019-12-06T14:57:19.5671247+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: read tcp <TRAFFIC_OPS_IP>:56546-><TRAFFIC_VAULT_IP>:8087: i/o timeout
ERROR: log.go:172: 2019-12-06T14:57:24.621252251+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: read tcp <TRAFFIC_OPS_IP>:56548-><TRAFFIC_VAULT_IP>:8087: i/o timeout
ERROR: log.go:172: 2019-12-06T14:57:29.672914624+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: read tcp <TRAFFIC_OPS_IP>:56550-><TRAFFIC_VAULT_IP>:8087: i/o timeout
ERROR: log.go:172: 2019-12-06T14:57:34.725151543+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: read tcp <TRAFFIC_OPS_IP>:56552-><TRAFFIC_VAULT_IP>:8087: i/o timeout
ERROR: log.go:172: 2019-12-06T14:57:39.776810812+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: read tcp <TRAFFIC_OPS_IP>:56554-><TRAFFIC_VAULT_IP>:8087: i/o timeout
ERROR: api.go:136: 2019-12-06T14:57:40.577628207+09:00: <TRAFFIC_ROUTER_IP>:35708 getting cdn ssl keys: getting cdn ssl keys from Riak: with cluster error: riak search error: executing Riak command index 'sslkeys' query 'cdn:kabletown_cdn': ClientError|[Cluster] all retries exhausted and/or no nodes available to execute command
```

#### Riak
```
2019-12-06 14:57:04.689 [error] <0.16781.1>@riak_api_pb_server:handle_info:331 Unrecognized message {error,closed}
2019-12-06 14:57:35.009 [error] <0.16889.1>@riak_api_pb_server:handle_info:331 Unrecognized message {error,closed}
2019-12-06 14:57:40.050 [error] <0.16912.1>@riak_api_pb_server:handle_info:331 Unrecognized message {error,closed}
2019-12-06 14:57:50.156 [error] <0.16945.1>@riak_api_pb_server:handle_info:331 Unrecognized message {error,closed}
2019-12-06 14:57:55.213 [error] <0.16963.1>@riak_api_pb_server:handle_info:331 Unrecognized message {error,closed}
```

### SSL disabled (`riak-admin security disable`)
#### Traffic Ops
```
ERROR: log.go:172: 2019-12-06T13:27:59.452543951+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: RiakError|0|Security not enabled; STARTTLS not allowed.
ERROR: log.go:172: 2019-12-06T13:28:04.45239435+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: RiakError|0|Security not enabled; STARTTLS not allowed.
ERROR: log.go:172: 2019-12-06T13:28:09.452478404+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: RiakError|0|Security not enabled; STARTTLS not allowed.
ERROR: +og.go:172: 2019-12-06T13:28:14.452369829+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: RiakError|0|Security not enabled; STARTTLS not allowed.
ERROR: log.go:172: 2019-12-06T13:28:19.452471924+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: RiakError|0|Security not enabled; STARTTLS not allowed.
ERROR: log.go:172: 2019-12-06T13:28:24.452359545+09:00: riak error: [ERROR] [Node] (<TRAFFIC_VAULT_IP>:8087|0|0) failed healthcheck in createConnection, err: RiakError|0|Security not enabled; STARTTLS not allowed.
ERROR: api.go:136: 2019-12-06T13:28:25.894461608+09:00: <TRAFFIC_ROUTER_IP>:34570 getting cdn ssl keys: getting cdn ssl keys from Riak: with cluster error: riak search error: executing Riak command index 'sslkeys' query 'cdn:kabletown_cdn': ClientError|[Cluster] all retries exhausted and/or no nodes available to execute command
``` 

#### Riak
no logs

**Comment by zrhoffman** [comment]
> > I recommend adding the following line to the bottom of your riak.conf, restarting, and trying again [per the riak docs](https://docs.riak.com/riak/kv/latest/using/security/basics/index.html#tls-settings):
> > ```
> > tls_protocols.tlsv1 = on
> > ```
> 
> I already add this line `tls_protocols.tlsv1 = on` same like in the guide. Yeah I know there is difference version curl in both OS. Maybe we need add more detail about traffic vault guide? So it the vault could be access by difference client OS.

I would suggest using `tls_protocols.tlsv1.1` instead of `tls_protocols.tlsv1`, which is TLS 1.0. Updating the Riak configuration is only half of the solution, because you also need to limit any client that connects to Riak to TLS 1.1:

```shell
curl --tlsv1.1 --tls-max 1.1 -vk https://trafficvault.ethernet.id:8088/types/default/props
```

> Can be accessed when requested from ubuntu 16.04 client.

This means that the OpenSSL library used by your version of `curl` in Ubuntu is not using any TLS signature algorithms that are unknown to Riak, which sidesteps the Erlang bug that @midchildan mentioned. You generally cannot configure this, because anywhere else, including signature algorithms unknown to the server is not an issue.

To demonstrate that this is really what is going on, compare the (successful) output of

```shell
openssl s_client -sigalgs 'RSA+SHA512' -connect trafficvault.ethernet.id:8088
```

vs. the unsuccessful output of

```shell
openssl s_client -sigalgs 'RSA+SHA512:RSA-PSS+SHA512' -connect trafficvault.ethernet.id:8088
```

**Comment by mitchell852** [comment]
closing as riak is soon to be deprecated.
